### PR TITLE
feat: OOO nickname change

### PR DESCRIPTION
**Pull request by bharati-21**
# Test: 1119 Update User Discord Nickname During OOO

### Closes: https://github.com/Real-Dev-Squad/website-backend/issues/1119

- The following PR contains changes related to the RDS discord server nickname change when a user is on OOO.
- The updates are reflected when the `PATCH /users/status/batch` API is triggered.

### Test Cases
- The tests for this PR are in #1460

### CODE REVIEW

File: services/usersStatusService.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    const response = await fetch(`${DISCORD_BASE_URL}/guild/member`, {
+      method: "PATCH",
+      headers: { "Content-Type": "application/json", Authorization: `Bearer ${authToken}` },
+      body: JSON.stringify({ discordId, userName }),
+    });
+    if (!response.ok) {
+      throw new Error("Could not update nickname");
+    }
+
+    await response.text();
```
Comment (heyrandhir) [inline services/usersStatusService.ts:17]: do we need this if we are not returning. we can remove it if we don't require it.

File: utils/userStatus.js [hunk 2]
```diff
@@ -302,6 +303,150 @@ const getNextDayTimeStamp = (timeStamp) => {
...
+  let oooMessage;
+
+  if (from && until) {
+    const fromDate = new Date(Number(from));
+    const fromDay = fromDate.getDate();
+    const fromMonth = month[fromDate.getMonth()];
+
+    oooMessage = `(OOO ${fromMonth} ${fromDay} - ${untilMonth} ${untilDay})`;
+  } else {
+    oooMessage = `(OOO ${untilMonth} ${untilDay})`;
```
Comment (heyrandhir) [inline utils/userStatus.js:327]: do we have a case where the user only gives until date and not from ?

File: utils/userStatus.js [hunk 3]
```diff
@@ -302,6 +303,150 @@ const getNextDayTimeStamp = (timeStamp) => {
   return nextDateDateTime.getTime();
 };
 
+/**
+ * Generates discord nickname for a user
+ *
+ * @param {string} username - The discord username of the user.
+ * @returns {string} - Nickname of the user.
+ */
+const generateOOONickname = (username, from, until) => {
```
Comment (heyrandhir) [inline utils/userStatus.js:312]: we can wrap the code inside the function in a try catch block

File: utils/userStatus.js [hunk 4]
```diff
@@ -302,6 +303,150 @@ const getNextDayTimeStamp = (timeStamp) => {
...
+  let errorsLen = 0;
+  try {
+    const updates = await Promise.allSettled(usersNicknameUpdates);
+    for (const update of updates) {
+      if (update.status === "rejected") {
+        errorsLen++;
+      }
+    }
+
+    if (errorsLen) {
```
Comment (heyrandhir) [inline utils/userStatus.js:368]: we are only sending the error metric. what about the success metric i.e how many user names were updated successfully.

**Review by heyrandhir — CHANGES_REQUESTED** [review]

File: services/usersStatusService.ts (hunk 1 above)
Comment (DashDeipayan) [inline services/usersStatusService.ts:17]: Should be returning something atleast, something like `{success:true}`

File: utils/userStatus.js [hunk 5]
```diff
@@ -302,6 +303,150 @@ const getNextDayTimeStamp = (timeStamp) => {
...
+  } catch (err) {
+    logger.error(`Failed to get discord id and username for user with id ${userId}`);
+    throw err;
+  }
+};
+
+const updateUsersDiscordNicknameBasedOnStatus = async (usersNicknameUpdates) => {
+  let errorsLen = 0;
+  try {
+    const updates = await Promise.allSettled(usersNicknameUpdates);
```
Comment (DashDeipayan) [inline utils/userStatus.js:361]: Why `Promise.allSettled()` instead of `Promise.all()`

**Review by DashDeipayan — CHANGES_REQUESTED** [review]

**Comment by bharati-21** [comment]
Closing thi

**Comment by bharati-21** [comment]
Closing this PR as a scheduled cron job needs to be created for the nickname change.

**Pull request closed without merging**